invenio_oaiserver/fetchers.py

```diff
@@ -33,3 +34,15 @@ def oaiid_fetcher(record_uuid, data):
         pid_type=OAIIDProvider.pid_type,
         pid_value=str(pid_value),
     )
+
+
+def set_records_query_fetcher(setSpec):
+    """Fetch query to retrieve records based on provided set spec."""
+    set = OAISet.query.filter(OAISet.spec == setSpec).first()
+    if set is None:
+        # raise error that no matches can be found ?
+        query = Q('match_none')
+    else:
+        query = Q(query_string_parser(set.search_pattern))
```

Comment:
What's happens if the `search_pattern` is empty? Why the default behaviour is not to search in the `_oai.set` fields?

Reply:
Thanks for your reply.

Here is some suggestions:

- [ ] If the `search_pattern` is required, I suggest to add a no null constraint in the db model
- [ ] to speed up the `ListIdentifiers`, you can use `source` for the elasticsearch during the `get_records`
- [ ] the source fields can be a configuration for `ListRecords`
- [ ] using a `search_query` as `insitution:test` raise the following error: `TypeError: expected string or bytes-like object`, replacing by `institution:"test"` works (bug)

Thanks

Reply:
> The idea is to not store the information of the sets a record belongs to within the record itself. Instead, the record is matched against the queries of the sets (ElasticSearch percolate query) and this information is generated dynamically on request.

Probably I misunderstand something. If during the request we use the `search_pattern`, is the percolator still useful?

Reply:
> Probably I misunderstand something. If during the request we use the `search_pattern`, is the percolator still useful?

The percolate query is used to get *all* sets a record belongs to (needed for the header of the response).
The additional query above is used, when a setSpec is specified in the request (f.e. https://127.0.0.1:5000/oai2d?verb=ListRecords&metadataPrefix=oai_dc&set=my-set). It is used during record querying so only records belonging to this set are retrieved. Then we take these records and do a percolate query to get all the sets the records belong to.

Reply:
@jma First of all thanks for reviewing! 

As @rekt-hard explained, the big change here is that we want to get away from storing OAI sets information inside the record. It's highly very inefficient to do it this way (e.g. define a new set, or modify an existing set and you may need to reindex every single record in the entire repository. The only thing we're loosing is the ability to say that a record was deleted from a set. A harvest can obtain that information by running ListIdentifiers for a set and compare against a local set.

On a side note, can I ask you not to use the "Request changes" when reviewing a PR. Instead, simply prefix your comment with Comment/Question/Minor/<nothing>/Major according (see definitions [here](https://invenio.readthedocs.io/en/latest/community/contributing/contribution-guide.html#pull-requests)). Basically idea is that lack of a green approve, means it's not approved. Just wanted to explain why I'm gonna press the "dismiss review"